### PR TITLE
Add support for 1D, 1D array, 2D array, 3D textures.

### PR DESCRIPTION
**Pull request by ben-clayton**
Fixes: #392

### CODE REVIEW

File: gapic/src/main/com/google/gapid/widgets/ImagePanel.java [hunk 1]
```diff
@@ -490,7 +490,7 @@ public Pixel getPixel(Point point) {
       float u = (float)imageNormalized.x * 0.5f + 0.5f;
       float v = (float)imageNormalized.y * (settings.flipped ? 0.5f : 0.5f) + 0.5f;
       int sampleY = settings.flipped ? (image.getHeight() - y - 1) : y;
-      return new Pixel(x, y, u, v, image.getData().getPixel(x, sampleY));
+      return new Pixel(x, y, u, v, image.getData().getPixel(x, sampleY, 1));
```
Comment (pmuetschard) [inline gapic/src/main/com/google/gapid/widgets/ImagePanel.java]: z should be zero here

File: gapic/src/main/com/google/gapid/image/ArrayImage.java [hunk 2]
```diff
@@ -35,25 +35,54 @@
...
+    int sliceSize = width * height * bytesPerPixel;
+    byte[] sliceData = Arrays.copyOfRange(data, sliceSize * z, sliceSize * (z + 1));
+    return create(width, height, 1, sliceData);
+  }
+
+  /**
+   * Constructs and returns a new {@link ArrayImage} of the same format with the given
+   * dimensions and data.
+   */
+  protected abstract ArrayImage create(int width, int height, int depth, byte[] data);
```
Comment (pmuetschard) [inline gapic/src/main/com/google/gapid/image/ArrayImage.java]: This should return Image.

File: gapic/src/main/com/google/gapid/image/Image.java [hunk 3]
```diff
@@ -66,57 +86,28 @@ public int getDepth() {
...
-      @Override
-      public void uploadToTexture(Texture texture) {
-        texture.loadData(0, 0, GL11.GL_RGB, GL11.GL_RGB, GL11.GL_UNSIGNED_BYTE, null);
-      }
-
-      @Override
-      public ImageData getImageData() {
-        return EMPTY_IMAGE;
-      }
+    public Image getSlice(int z) { return this; }
```
Comment (pmuetschard) [inline gapic/src/main/com/google/gapid/image/Image.java]: Please add newlines after { and before }

File: gapic/src/main/com/google/gapid/glviewer/vec/VecD.java [hunk 4]
```diff
@@ -139,20 +139,28 @@ public VecD divide(double sx, double sy, double sz) {
...
-    return x * x + y * y + z * z;
+  /** Like {@link #divide(VecD)}, but divide by zeros result in zero. */
+  public VecD safeDivide(VecD v) {
+    return safeDivide(v.x, v.y, v.z);
   }
 
+  public double magnitudeSquared() {
+  return x * x + y * y + z * z;
+}
+
```
Comment (pmuetschard) [inline gapic/src/main/com/google/gapid/glviewer/vec/VecD.java]: undo

File: gapic/src/main/com/google/gapid/image/FetchedImage.java [hunk 5]
```diff
@@ -94,10 +94,12 @@
...
 
-  public static ListenableFuture<ImageData> loadLevel(
-      ListenableFuture<FetchedImage> futureImage, final int level) {
+  public static ListenableFuture<ImageData> loadImage(
+      ListenableFuture<FetchedImage> futureImage, final int layer, final int level) {
     return Futures.transformAsync(futureImage, image -> Futures.transform(
-        image.getLevel(Math.min(level, image.getLevelCount())), (l) -> l.getImageData()));
+        image.getImage(
+            Math.min(layer, image.getLayerCount()),
+            Math.min(level, image.getLevelCount())), (l) -> l.getImageData()));
```
Comment (pmuetschard) [inline gapic/src/main/com/google/gapid/image/FetchedImage.java]: `getLayerCount() - 1` and `getLevelCount() - 1`, no?

File: gapic/src/main/com/google/gapid/widgets/ImagePanel.java [hunk 6]
```diff
@@ -649,11 +692,12 @@ private void onScroll() {
     private static final int PREVIEW_HEIGHT = 11; // Should be odd, so center pixel looks nice.
     private static final int PREVIEW_SIZE = 7;
 
+
+    private final Map<Image, Texture> imageToTexture = new HashMap<>();
```
Comment (pmuetschard) [inline gapic/src/main/com/google/gapid/widgets/ImagePanel.java]: `Maps.newHashMap()` for consistency

File: gapic/src/main/com/google/gapid/widgets/ImagePanel.java [hunk 7]
```diff
@@ -672,25 +716,37 @@ public void init(Renderer renderer) {
...
     @Override
     public void update(Renderer renderer, SceneData newData) {
-      Image oldImage = (this.data != null) ? this.data.image : null;
-      if (oldImage != newData.image) {
-        // Release old texture, create new.
-        if (texture != null) {
-          texture.delete();
+      // Release textures that are no longer in data.
+      Set<Image> newSet = Sets.newHashSet(newData.images);
+      for (Entry<Image, Texture> entry : imageToTexture.entrySet()) {
```
Comment (pmuetschard) [inline gapic/src/main/com/google/gapid/widgets/ImagePanel.java]: nit: use `Map.Entry` and remove the import. `Entry` is so generic and `Map.Entry` so short, that I prefer it that way.

File: gapic/src/main/com/google/gapid/image/ArrayImageBuffer.java [hunk 8]
```diff
@@ -87,25 +88,34 @@ protected static ByteBuffer buffer(byte[] data) {
...
-      } else {
-        // Copy one (incomplete) row at a time.
-        for (int row = 0, p = y * width, s = 0; row < h; row++, p += width, s += w * pixelSize) {
-          System.arraycopy(src, s, data, pixelSize * (p + x), pixelSize * w);
+    public Builder update(byte[] src, int x, int y, int z, int w, int h, int d) {
+      if (x == 0 && y == 0 && w == width && h == height) {
+        // Bulk copy.
+        System.arraycopy(src, 0, data, pixelSize * w * h * z, pixelSize * w * h * d);
+      }
+      for (int slice = 0; slice < d; slice++) {
```
[the remark this answers is not in the document]
Comment (ben-clayton) [reply gapic/src/main/com/google/gapid/image/ArrayImageBuffer.java]: Done

File: gapic/src/main/com/google/gapid/image/ArrayImageBuffer.java [hunk 9]
```diff
@@ -87,25 +88,34 @@ protected static ByteBuffer buffer(byte[] data) {
...
+        int srcOffset = pixelSize * slice * width * height;
+        int dstOffset = pixelSize * slice * w * h;
```
[the remark this answers is not in the document]
Comment (ben-clayton) [reply gapic/src/main/com/google/gapid/image/ArrayImageBuffer.java]: Done

File: gapic/src/main/com/google/gapid/widgets/ImagePanel.java (hunk 1 above)
Comment (ben-clayton) [reply gapic/src/main/com/google/gapid/widgets/ImagePanel.java]: Good spot. Done.

File: gapic/src/main/com/google/gapid/widgets/ImagePanel.java (hunk 6 above)
Comment (ben-clayton) [reply gapic/src/main/com/google/gapid/widgets/ImagePanel.java]: Done

File: gapic/src/main/com/google/gapid/widgets/ImagePanel.java (hunk 7 above)
Comment (ben-clayton) [reply gapic/src/main/com/google/gapid/widgets/ImagePanel.java]: Done

File: gapic/src/main/com/google/gapid/image/ArrayImage.java (hunk 2 above)
Comment (ben-clayton) [reply gapic/src/main/com/google/gapid/image/ArrayImage.java]: Done

File: gapic/src/main/com/google/gapid/image/FetchedImage.java (hunk 5 above)
Comment (ben-clayton) [reply gapic/src/main/com/google/gapid/image/FetchedImage.java]: Good spot - bug was carried over from before this PR.

File: gapic/src/main/com/google/gapid/glviewer/vec/VecD.java (hunk 4 above)
Comment (ben-clayton) [reply gapic/src/main/com/google/gapid/glviewer/vec/VecD.java]: Done

File: gapic/src/main/com/google/gapid/image/Image.java (hunk 3 above)
Comment (ben-clayton) [reply gapic/src/main/com/google/gapid/image/Image.java]: Done

**Comment by Qining** [comment]
I have a CL to extend this to the Vulkan side, will be posted after this is in.

**Pull request merged**